### ISSUE TITLE
Control the starting seed?

### ISSUE THREAD

**Issue by dginev** [lucidrains/deep-daze#9]
I wanted to make a PR, but I couldn't figure out where would be the best place to intervene in the cascade of initialization calls upto Siren itself.

Compared to the original notebook, which randomizes its initial seed on each run, the `deep-daze` approach currently always seems to start with a very noir seed:
<img src="https://user-images.githubusercontent.com/348975/104987090-d2db4b00-59e2-11eb-9a07-6c7ffe018e0c.png" width=250>

and amazingly CLIP is quite happy to navigate inside it and keep generating in that style (even when I would prefer a cheery light generation instead). Here are two examples of the types of generation one ends up with:
<table><tr><td>
<img src="https://user-images.githubusercontent.com/348975/104987169-07e79d80-59e3-11eb-9244-0ae2c750a25a.png" width=250> </td><td>
<img src="https://user-images.githubusercontent.com/348975/104987181-12a23280-59e3-11eb-9ee1-40026627ae0b.png" width=250></td></tr></table>

Ideally one would like to specify some hex color as "dominant" for the seed and expose that as a command line option, but I could do that myself if I found out where to look. Could you ( @lucidrains ) point me to the right general area where one would set a custom init seed for Siren?

**Comment by lucidrains** [comment]
@dginev Hi Deyan again! I took a look at the code and realized it was probably because of this extra tanh I introduced to the output of the Siren. I have since removed it from the most recent versions and generations seems to exhibit more range of brightness from the very beginning. Do you want to give it another run and see if that works for you?

**Comment by dginev** [comment]
Just to be sure I did:
```
pip install --upgrade deep-daze --force
```

and I'm seeing the same fixed seed. Wondering if I should wipe my `~/.cache` alongside... But yeah, for now I'm still stuck with that seed, good to hear I may just need a clean setup to escape it.

**Comment by lucidrains** [comment]
@dginev I don't think the seed is the problem, but since you brought it up, i added the ability to specify it with `--seed` in the latest version

**Comment by dginev** [comment]
You're right, that changed nothing for me, same initial image whatever I give to `--seed`. I'll sleep on it and tinker some more myself tomorrow. Something odd must be going on with my installation if I'm the only one seeing this.

**Comment by lucidrains** [comment]
@dginev I am just as in the dark as you are (no pun intended) i've been examining this line https://github.com/lucidrains/deep-daze/blob/main/deep_daze/deep_daze.py#L226 and think the normalization here is unnecessary

however, this also exists in the original colab repo

i'll have to think about it some more as well

**Comment by afiaka87** [comment]
I think you need to do the "inverse" normalization as the model will only generate content close to the normalization it was trained on. Here's the code that does that, sorry I'm busy and haven't been able to test it. Just jotted it down in the GitHub UI. 

https://github.com/lucidrains/deep-daze/pull/11

Could be way off base. This is the first time I've actually looked at ML code.

**Comment by afiaka87** [comment]
## Long version
From CLIP's [Colab notebook](https://github.com/openai/CLIP/blob/6bc0bd88737ac3b28a5b3288fdd13b7a6cd94b22/Interacting_with_CLIP.ipynb):
> We resize the input images and center-crop them **to conform with the image resolution that the model expects**. Before doing so, we will normalize the pixel intensity using **the dataset mean and standard deviation.**
```python
preprocess = Compose([
    Resize(input_resolution, interpolation=Image.BICUBIC),
    CenterCrop(input_resolution),
    ToTensor()
])
image_mean = torch.tensor([0.48145466, 0.4578275, 0.40821073]).cuda()
image_std = torch.tensor([0.26862954, 0.26130258, 0.27577711]).cuda()
# Several lines later ...
image_input = torch.tensor(np.stack(images)).cuda()
image_input -= image_mean[:, None, None]
image_input /= image_std[:, None, None]
```


### And here's the transform 
```python
  transform = Compose([
      Resize(n_px, interpolation=Image.BICUBIC),
      CenterCrop(n_px),
      lambda image: image.convert("RGB"),
      ToTensor(),
      Normalize((0.48145466, 0.4578275, 0.40821073), (0.26862954, 0.26130258, 0.27577711)),
  ])
````

So here's where you're using it 
https://github.com/lucidrains/deep-daze/blob/89fb3a2e4bd33c7534a6a9cf219b8974c603e436/deep_daze/clip.py#L169
https://github.com/lucidrains/deep-daze/blob/89fb3a2e4bd33c7534a6a9cf219b8974c603e436/deep_daze/deep_daze.py#L226
                img = normalize_image(self.model(self.encoded_text, return_loss = False).cpu())

**Comment by afiaka87** [comment]
https://github.com/lucidrains/deep-daze/issues/9#issuecomment-762937171

This is the result of that change.  Has good detail, but the contrast is awful. I'm stumped. 
![an_apple_next_to_a_fireplace](https://user-images.githubusercontent.com/3994972/105071087-c0e0c300-5a49-11eb-9be7-20d9902f85ad.png)

**Comment by dginev** [comment]
Just to be clear I'm understanding this right @afiaka87 , can you also reproduce getting the dark initial image I posted in the description of the issue on every run of the released `deep-daze` ? Or the `main` branch for that matter. If I'm not alone I will be debugging this a little differently.

**Comment by afiaka87** [comment]
@dginev Yes, my Colab instances are consistently starting with a "dark blue noise" every time. 
@lucidrains Sorry, I'm gonna be booting up my actual linux partition so I can look deeper into this, but I'm getting much better output on an initial test with this change https://github.com/afiaka87/deep-daze/pull/1

**Comment by lucidrains** [comment]
@dginev hi Deyan, I realized the seed wasn't done correctly when you tried it lol

Do you want to give it another go?

**Comment by dginev** [comment]
deep-daze 0.2.9
```
$ imagine --seed=94022 --num-layers 32 "seed of hope"
```
<img width=250 src="https://user-images.githubusercontent.com/348975/105110522-c3184100-5a8c-11eb-9cde-d97050870d46.png">

```
$ imagine --seed=10022 --num-layers 32 "seed of hope"
```

<img width=250 src="https://user-images.githubusercontent.com/348975/105110670-1094ae00-5a8d-11eb-92e0-dba6ee925171.png">

```
$ imagine --seed=666666 --num-layers 32 "seed of hope"
```

<img width=250 src="https://user-images.githubusercontent.com/348975/105110728-4043b600-5a8d-11eb-8a32-42bb43e930fa.png">

```
imagine --seed=666999666 --num-layers 32 "seed of hope"
```

<img width=250 src="https://user-images.githubusercontent.com/348975/105110827-813bca80-5a8d-11eb-9581-6bf842163b45.png">

---

It has _some_ effect now! But it's subtle? Two wrongs didn't make a right this time :>

**Comment by dginev** [comment]
FYI, I tried to see if the subtle differences aren't enough for the generation to escape the "perpetual negative space" it's seeded in, but it doesn't really manage - which continues to amaze me. Here's "seed of hope" after a thousand steps, `--seed=872072`:
<img width=250 src="https://user-images.githubusercontent.com/348975/105111978-0f18b500-5a90-11eb-9e20-13b67fa7570c.png">

**Comment by afiaka87** [comment]
I'm definitely still having issues with it getting too hung up on the intial blue noise. 

I'm out of my depth at this point. But here's a comparison of 2000 iterations on the phrase "angry protestors". First one is normalized as per @lucidrains 
![normalized](https://user-images.githubusercontent.com/3994972/105115104-d972cc80-5a8d-11eb-9d02-9639444dd584.png)

Second one doesn't get a normalization before it's saved as per https://github.com/afiaka87/deep-daze/pull/1
![denormalized](https://user-images.githubusercontent.com/3994972/105115110-dd065380-5a8d-11eb-98a1-39f7e750b162.png)

Didn't specify the day at all, but yeah my approach is basically having the same problem but with "earthiness" instead of the "noir" that the blue creates.

**Comment by lucidrains** [comment]
maybe we can tweak the normalization for the siren output before the image is saved until it is satisfactory for a set of images of different colors?

**Comment by afiaka87** [comment]
Yeah I think the colors can be corrected but it's beyond me at the moment. Still trying to understand this random slicing code ha. Pretty cool stuff. But yeah it's just heavily dependent on prior input at the moment. It's possible that predicting out two or three more frames and having CLIP choose the correct one could help it to diverge more.

**Comment by afiaka87** [comment]
Some old school image manipulation could probably come in handy. Introducing random noise into the image from time to time would definitely force it to diverge. Could parameterize it perhaps.

**Comment by dginev** [comment]
My thinking when I opened this issue was - "hey this worked really great in the notebook, which drew a random seed each time, usually in a different color than the previous one."

So I'll try to figure out how to get that effect back in...

**Comment by afiaka87** [comment]
Well in the meantime, I made a barebones version of the old one that's easier to run. Just three cells. 

https://github.com/afiaka87/deep-daze/blob/main/Notebook_Easy.ipynb

**Comment by torridgristle** [comment]
By rendering Kramer in various prompts, I've found that using
```
def norm(normthis):
  maxthis = torch.amax(normthis)
  minthis = torch.amin(normthis)
  return torch.divide(torch.subtract(normthis, minthis),maxthis-minthis)
```
works just fine for the output image, but with the usual "nom" using the mean and stdev of the dataset CLIP was trained on still in use for the ascend_txt function.

**Comment by afiaka87** [comment]
@torridgristle I believe I implemented what you're referring to in this PR https://github.com/lucidrains/deep-daze/pull/14

I'll start testing it out now. Lemme know if I didn't understand you.

**Comment by afiaka87** [comment]
Here's a colab notebook with the changes (just pip installing my specific branch) if you want to try it your self 
https://github.com/afiaka87/deep-daze/blob/torrid_suggestion/Deep_Daze_%2B_Torrid_Suggestions.ipynb

**Comment by lucidrains** [comment]
@afiaka87 i'll try playing with the SIREN initial bias values next week

**Comment by torridgristle** [comment]
Would a gamma adjustment when generating the image to be put through CLIP be beneficial? I'm wondering if it looks washed out because it's generating images without being darkened, or perhaps without being brightened. I'd try it myself but the dang thing keeps giving me errors about the a minimum value not being finite or something whenever I try to pow what I believe to be the image after it's resized to 224.

**Comment by afiaka87** [comment]
While I am curious why the results here are different than the original notebook, I'll just say that I'm not exactly sure this is a "bug" per se. Given enough layers, iterations and epochs, generated images will eventually escape their blue hue. It does take quite awhile, but perhaps that's the nature of CLIP not caring what the hue of an image. 

I tried to reverse engineer the differences between @lucidrains repo and the original notebook and, well, it seems to be mostly the same. Even the apparent loss of the `mgrid` function mentioned in another issue was actually just inside the cloned SIREN code. And I think this repo's code is far more legible as well (thank you so much for figuring out what all those single/double/triple letter variable names were actually referring to). 

Anyway, really the only thing i'm concerned about is that the two notebooks output different values. Knowing why that is would be useful, but I'm no longer convinced that this "blue noise" is inherently the problem. The nice thing about this model is you can literally just say things like "in daytime" to help get it out of that specific rut. And again, I've found that the blue hue seems to entirely disappear given more than a few thousand iterations.

**Comment by dginev** [comment]
>  I'll just say that I'm not exactly sure this is a "bug" per se. Given enough layers, iterations and epochs, generated images will eventually escape their blue hue. It does take quite awhile, but perhaps that's the nature of CLIP not caring what the hue of an image.

I'm happy to keep claiming this is preventing me from using the CLI tool, because I expect to have a rough idea of what the saturated final image will be around 1000 iterations in, when using the original notebook. Having to wait 10,000 steps to (maybe) escape the initial seed framing can definitely be claimed as unusable, since the notebook does exist and will give me better results for a lot less compute.

**Comment by afiaka87** [comment]
> > I'll just say that I'm not exactly sure this is a "bug" per se. Given enough layers, iterations and epochs, generated images will eventually escape their blue hue. It does take quite awhile, but perhaps that's the nature of CLIP not caring what the hue of an image.
> 
> I'm happy to keep claiming this is preventing me from using the CLI tool, because I expect to have a rough idea of what the saturated final image will be around 1000 iterations in, when using the original notebook. Having to wait 10,000 steps to (maybe) escape the initial seed framing can definitely be claimed as unusable, since the notebook does exist and will give me better results for a lot less compute.

That's fair. My experience has been a little more like "it tends to escape the blues after around iteration 400 with 32 hidden layers and a batch_size of 16". This however, isn't the case on certain input that is inherently rewarded for being blue for whatever reason. 

At this point, I'm still digging around in the code looking for where the difference between the original notebook and the current one is but I'm not exactly an expert. In the meantime, I guess you might try changing up your phrasing? "In daytime" or even maybe "with a warm hue" could be helpful to append to your phrases.

**Comment by dginev** [comment]
Now that `start_image_path` exists, I can escape into  "happier" latent space! So, as far as I'm concerned, this issue can be closed and my solution is to use `--start_image_path` with some appropriate seed image.

One interesting peculiarity about that option, it seems that it works "as expected" only if the number of layers is left untouched (i.e. at 16) - if I go to 32 layers, the seed image is completely distorted right off the bat - but maybe there's a way to tweak the parameters to avoid that. In any case, works for me - thanks @lucidrains

**Comment by afiaka87** [comment]
@dginev I think we can close this issue now. I'm still sort of just generally curious about why the original research notebook gives a different initial color on the same value of theta (30.0) than this repo. But I believe the issue here covers too wide a range of discussion.

**Comment by afiaka87** [comment]
@dginev I believe @NotNANtoN has finally fixed this problem with his latest pull requst!
